{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# PyMol Mutation Visualization\n",
    "\n",
    "This notebooks can be used to visualize mutation statistics from cBioPortal in PyMol. First mutation statistics will analyzed. Then a text file will be generated whose content can be used in PyMol."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "ntrks_data = {\n",
    "    \"ntrk1\": {\n",
    "        \"input_path\": \"../../data/NTRK1_lollipop.tsv\",\n",
    "        \"output_path\": \"../../data/NTRK1_pymol_commands\",\n",
    "        \"start\": 510,\n",
    "        \"end\": 781},\n",
    "    \"ntrk2\": {\n",
    "        \"input_path\": \"../../data/NTRK2_lollipop.tsv\",\n",
    "        \"output_path\": \"../../data/NTRK2_pymol_commands\",\n",
    "        \"start\": 538,\n",
    "        \"end\": 807},\n",
    "    \"ntrk3\": {\n",
    "        \"input_path\": \"../../data/NTRK3_lollipop.tsv\",\n",
    "        \"output_path\": \"../../data/NTRK3_pymol_commands\",\n",
    "        \"start\": 538,\n",
    "        \"end\": 839}\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Study</th>\n",
       "      <th>Sample ID</th>\n",
       "      <th>Cancer Type</th>\n",
       "      <th>Protein Change</th>\n",
       "      <th>Annotation</th>\n",
       "      <th>Functional Impact</th>\n",
       "      <th>Mutation Type</th>\n",
       "      <th>Variant Type</th>\n",
       "      <th>Copy #</th>\n",
       "      <th>COSMIC</th>\n",
       "      <th>...</th>\n",
       "      <th>Allele Freq (N)</th>\n",
       "      <th>Variant Reads</th>\n",
       "      <th>Ref Reads</th>\n",
       "      <th>Variant Reads (Normal)</th>\n",
       "      <th>Ref Reads (Normal)</th>\n",
       "      <th># Mut in Sample</th>\n",
       "      <th>Exon</th>\n",
       "      <th>gnomAD</th>\n",
       "      <th>ClinVar ID</th>\n",
       "      <th>dbSNP</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Sarcoma (TCGA, PanCancer Atlas)</td>\n",
       "      <td>TCGA-DX-A3UA-01</td>\n",
       "      <td>Leiomyosarcoma</td>\n",
       "      <td>TPM3-NTRK1</td>\n",
       "      <td>OncoKB: Likely Oncogenic, level_1;CIViC: NA;My...</td>\n",
       "      <td>MutationAssessor: NA;SIFT: NA;Polyphen-2: NA</td>\n",
       "      <td>Fusion</td>\n",
       "      <td>NaN</td>\n",
       "      <td>diploid</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>26.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Thyroid Carcinoma (TCGA, PanCancer Atlas)</td>\n",
       "      <td>TCGA-EL-A3D4-01</td>\n",
       "      <td>Papillary Thyroid Cancer</td>\n",
       "      <td>TPM3-NTRK1</td>\n",
       "      <td>OncoKB: Likely Oncogenic, level_1;CIViC: NA;My...</td>\n",
       "      <td>MutationAssessor: NA;SIFT: NA;Polyphen-2: NA</td>\n",
       "      <td>Fusion</td>\n",
       "      <td>NaN</td>\n",
       "      <td>diploid</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>6.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Bladder Urothelial Carcinoma (TCGA, PanCancer ...</td>\n",
       "      <td>TCGA-FD-A6TI-01</td>\n",
       "      <td>Bladder Urothelial Carcinoma</td>\n",
       "      <td>TPM3-NTRK1</td>\n",
       "      <td>OncoKB: Likely Oncogenic, level_1;CIViC: NA;My...</td>\n",
       "      <td>MutationAssessor: NA;SIFT: NA;Polyphen-2: NA</td>\n",
       "      <td>Fusion</td>\n",
       "      <td>NaN</td>\n",
       "      <td>gain</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>115.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>MSK-IMPACT Clinical Sequencing Cohort (MSKCC, ...</td>\n",
       "      <td>P-0003329-T01-IM5</td>\n",
       "      <td>Gallbladder Cancer</td>\n",
       "      <td>NTRK1-LMNA fusion</td>\n",
       "      <td>OncoKB: Oncogenic, level_1;CIViC: NA;MyCancerG...</td>\n",
       "      <td>MutationAssessor: NA;SIFT: NA;Polyphen-2: NA</td>\n",
       "      <td>Fusion</td>\n",
       "      <td>NaN</td>\n",
       "      <td>diploid</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>MSK-IMPACT Clinical Sequencing Cohort (MSKCC, ...</td>\n",
       "      <td>P-0003705-T01-IM5</td>\n",
       "      <td>Mucinous Adenocarcinoma of the Appendix</td>\n",
       "      <td>NTRK1-LMNA fusion</td>\n",
       "      <td>OncoKB: Oncogenic, level_1;CIViC: NA;MyCancerG...</td>\n",
       "      <td>MutationAssessor: NA;SIFT: NA;Polyphen-2: NA</td>\n",
       "      <td>Fusion</td>\n",
       "      <td>NaN</td>\n",
       "      <td>diploid</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows Ã— 31 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                               Study          Sample ID  \\\n",
       "0                    Sarcoma (TCGA, PanCancer Atlas)    TCGA-DX-A3UA-01   \n",
       "1          Thyroid Carcinoma (TCGA, PanCancer Atlas)    TCGA-EL-A3D4-01   \n",
       "2  Bladder Urothelial Carcinoma (TCGA, PanCancer ...    TCGA-FD-A6TI-01   \n",
       "3  MSK-IMPACT Clinical Sequencing Cohort (MSKCC, ...  P-0003329-T01-IM5   \n",
       "4  MSK-IMPACT Clinical Sequencing Cohort (MSKCC, ...  P-0003705-T01-IM5   \n",
       "\n",
       "                               Cancer Type     Protein Change  \\\n",
       "0                           Leiomyosarcoma         TPM3-NTRK1   \n",
       "1                 Papillary Thyroid Cancer         TPM3-NTRK1   \n",
       "2             Bladder Urothelial Carcinoma         TPM3-NTRK1   \n",
       "3                       Gallbladder Cancer  NTRK1-LMNA fusion   \n",
       "4  Mucinous Adenocarcinoma of the Appendix  NTRK1-LMNA fusion   \n",
       "\n",
       "                                          Annotation  \\\n",
       "0  OncoKB: Likely Oncogenic, level_1;CIViC: NA;My...   \n",
       "1  OncoKB: Likely Oncogenic, level_1;CIViC: NA;My...   \n",
       "2  OncoKB: Likely Oncogenic, level_1;CIViC: NA;My...   \n",
       "3  OncoKB: Oncogenic, level_1;CIViC: NA;MyCancerG...   \n",
       "4  OncoKB: Oncogenic, level_1;CIViC: NA;MyCancerG...   \n",
       "\n",
       "                              Functional Impact Mutation Type Variant Type  \\\n",
       "0  MutationAssessor: NA;SIFT: NA;Polyphen-2: NA        Fusion          NaN   \n",
       "1  MutationAssessor: NA;SIFT: NA;Polyphen-2: NA        Fusion          NaN   \n",
       "2  MutationAssessor: NA;SIFT: NA;Polyphen-2: NA        Fusion          NaN   \n",
       "3  MutationAssessor: NA;SIFT: NA;Polyphen-2: NA        Fusion          NaN   \n",
       "4  MutationAssessor: NA;SIFT: NA;Polyphen-2: NA        Fusion          NaN   \n",
       "\n",
       "    Copy #  COSMIC  ... Allele Freq (N) Variant Reads Ref Reads  \\\n",
       "0  diploid     NaN  ...             NaN           NaN       NaN   \n",
       "1  diploid     NaN  ...             NaN           NaN       NaN   \n",
       "2     gain     NaN  ...             NaN           NaN       NaN   \n",
       "3  diploid     NaN  ...             NaN           NaN       NaN   \n",
       "4  diploid     NaN  ...             NaN           NaN       NaN   \n",
       "\n",
       "   Variant Reads (Normal)  Ref Reads (Normal)  # Mut in Sample Exon gnomAD  \\\n",
       "0                     NaN                 NaN             26.0  NaN    NaN   \n",
       "1                     NaN                 NaN              6.0  NaN    NaN   \n",
       "2                     NaN                 NaN            115.0  NaN    NaN   \n",
       "3                     NaN                 NaN              1.0  NaN    NaN   \n",
       "4                     NaN                 NaN              3.0  NaN    NaN   \n",
       "\n",
       "  ClinVar ID dbSNP  \n",
       "0        NaN   NaN  \n",
       "1        NaN   NaN  \n",
       "2        NaN   NaN  \n",
       "3        NaN   NaN  \n",
       "4        NaN   NaN  \n",
       "\n",
       "[5 rows x 31 columns]"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mutations = pd.read_csv(ntrks_data[\"ntrk1\"][\"input_path\"], sep=\"\\t\")\n",
    "mutations.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['G595R', 'G595W', 'G595W', 'T264M', 'T264M']\n",
      "Number of missense mutations: 594\n"
     ]
    }
   ],
   "source": [
    "missense_mutations = mutations[mutations[\"Mutation Type\"] == \"Missense_Mutation\"][\"Protein Change\"].to_list()\n",
    "print(missense_mutations[:5])\n",
    "print(f\"Number of missense mutations: {len(missense_mutations)}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_missense_mutations_dict(missense_mutations, start, end):\n",
    "    missense_mutations_dict = {}\n",
    "    for mutation in missense_mutations:\n",
    "        try:\n",
    "            resid = int(mutation[1:-1])\n",
    "        except ValueError as e:\n",
    "            print(e)\n",
    "            continue\n",
    "        residue_new = mutation[-1]\n",
    "        if start <= resid <= end:\n",
    "            if resid not in missense_mutations_dict.keys():\n",
    "                missense_mutations_dict[resid] = {\"count\": 1, \"amino_acids\": set(residue_new)}\n",
    "            else:\n",
    "                missense_mutations_dict[resid][\"count\"] += 1\n",
    "                missense_mutations_dict[resid][\"amino_acids\"].add(residue_new)\n",
    "    return missense_mutations_dict"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_sphere_scale(number):\n",
    "    min_sphere_scale = 0.5\n",
    "    max_sphere_scale = 1.5\n",
    "    sphere_scale = min_sphere_scale + (number / 10)\n",
    "    if sphere_scale > max_sphere_scale:\n",
    "        return max_sphere_scale\n",
    "    else:\n",
    "        return sphere_scale"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "def write_pymol_visualization(output_path, missense_mutations_dict):\n",
    "    with open(output_path, \"w\") as wf:\n",
    "        for resid, data in missense_mutations_dict.items():\n",
    "            wf.write(f\"create {resid}, resid {resid} and name CA\\n\")\n",
    "            wf.write(f\"show_as sphere, {resid}\\n\")\n",
    "            wf.write(f\"set sphere_scale, {get_sphere_scale(data['count'])}, {resid}\\n\")\n",
    "            wf.write(f\"color 0x9e000d, {resid}\\n\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Analyzing mutations for ntrk1 ...\n",
      "Number of missense mutations: 594\n",
      "invalid literal for int() with base 10: '216_V217delinsH'\n",
      "invalid literal for int() with base 10: '216_V217delinsH'\n",
      "invalid literal for int() with base 10: '216_V217delinsH'\n",
      "Number of mutated residues in kinase domain: 109\n",
      "Number of unique mutations in kinase domain: 149\n",
      "Total number of observed mutations: 211\n",
      "Writing pymol visulaization script ...\n",
      "Analyzing mutations for ntrk2 ...\n",
      "Number of missense mutations: 459\n",
      "Number of mutated residues in kinase domain: 116\n",
      "Number of unique mutations in kinase domain: 138\n",
      "Total number of observed mutations: 182\n",
      "Writing pymol visulaization script ...\n",
      "Analyzing mutations for ntrk3 ...\n",
      "Number of missense mutations: 835\n",
      "Number of mutated residues in kinase domain: 166\n",
      "Number of unique mutations in kinase domain: 224\n",
      "Total number of observed mutations: 325\n",
      "Writing pymol visulaization script ...\n"
     ]
    }
   ],
   "source": [
    "for ntrk, data in ntrks_data.items():\n",
    "    print(f\"Analyzing mutations for {ntrk} ...\")\n",
    "    mutations = pd.read_csv(data[\"input_path\"], sep=\"\\t\")\n",
    "    missense_mutations = mutations[mutations[\"Mutation Type\"] == \"Missense_Mutation\"][\"Protein Change\"].to_list()\n",
    "    print(f\"Number of missense mutations: {len(missense_mutations)}\")\n",
    "    missense_mutations_dict = get_missense_mutations_dict(missense_mutations, data[\"start\"], data[\"end\"])\n",
    "    print(f\"Number of mutated residues in kinase domain: {len(missense_mutations_dict)}\")\n",
    "    print(f\"Number of unique mutations in kinase domain: {sum([len(value['amino_acids']) for value in missense_mutations_dict.values()])}\")\n",
    "    print(f\"Total number of observed mutations: {sum([value['count'] for value in missense_mutations_dict.values()])}\")\n",
    "    print(\"Writing pymol visulaization script ...\")\n",
    "    write_pymol_visualization(data[\"output_path\"], missense_mutations_dict)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
